examples/run_nlp_trainer.py

```diff
@@ -0,0 +1,373 @@
+        tb.append(output_line)
+    
+    print(tabulate(tb, headers='firstrow'))
+
+''' Callback function for metric collection. '''
+def log_metrics(parent_ckpt: Optional[str], ckpt_filename: str, glue_metrics: GlueState, f_metric: Future) -> None:
+    metric = f_metric.result()
+    if parent_ckpt: 
+        logged_ckpt_name = parent_ckpt
+    else:
+        logged_ckpt_name = ckpt_filename
+
+    if logged_ckpt_name not in glue_metrics.get_logged_ckpts():
+        glue_metrics.init_ckpt_dict(logged_ckpt_name)
+            
+    task = list(metric.keys())[0]
+    formatted_task = task.split('_')[1] # remove "glue_" prefix
+    for key in metric.keys(): # handle case where mnli has glue_mnli and glue_mnli_mismatched
+        for m in metric[key]:
+            if not glue_metrics.get_task_metric(logged_ckpt_name, formatted_task):
+                glue_metrics.set_task_metric(logged_ckpt_name, formatted_task, metric[key][m].item())
+            else:
+                old_val = glue_metrics.get_task_metric(logged_ckpt_name, formatted_task)
+                glue_metrics.set_task_metric(logged_ckpt_name, formatted_task, sum([old_val, metric[key][m].item()]) / 2) # automatically average scores 
+    
+''' Set up CUDA environment and process pool for given finetuning jobs. '''
+def setup_finetuning_jobs(finetune_tasks: List[str], save_ckpts: List[bool], ckpt_load_path: str, ckpt_filename: str, save_folder: str, glue_metrics: GlueState,
+         wandb_logger: Union[WandBLogger, None], save_locally: bool, bucket: Optional[str] = None, parent_ckpt: Optional[str] = None, load_ignore_keys: Optional[List[str]] = None) -> None:
+    cuda_envs = init_cuda_queue(torch.cuda.device_count())
+    num_tasks = len(finetune_tasks)
+    
+    if parent_ckpt: 
+        wandb_group_name = parent_ckpt
+    else:
+        wandb_group_name = ckpt_load_path
+
+    # finetuning from pretrained checkpoint
+    print(f"FINETUNING ON {ckpt_filename}!")
+    futures = []
+    log_metrics_partial = partial(log_metrics, parent_ckpt, ckpt_filename, glue_metrics)
+    executor = Pool(max_workers=torch.cuda.device_count(), initializer=init_cuda_env, initargs=(cuda_envs, ))
+    for rank in range(num_tasks):
+        future = executor.submit(train_finetune, finetune_tasks[rank], save_ckpts[rank], ckpt_load_path, wandb_group_name, save_folder, save_locally, wandb_logger, bucket, load_ignore_keys)
+        future.add_done_callback(log_metrics_partial)
+        # future.add_done_callback(log_metrics)
+        futures.append(future)
+
+    executor.shutdown(wait=True) # wait for processes and callbacks to complete 
+  
+    cuda_envs.close() 
+    cuda_envs.join_thread()
+    
+'''Run single instance of a finetuning job on given task. ''' 
+def train_finetune(task: str, save_ckpt: bool, load_path: str, wandb_group_name: str, save_folder: str, save_locally: bool, wandb_logger: Union[WandBLogger, None], bucket: Optional[str] = None, load_ignore_keys: Optional[List[str]] = None):
```

Comment:
Can we remove the `wandb` and `bucket` arguments here, and instead require that the user specify wandb and/or the object store logger as part of their yaml?

Reply:
Right now, the user has to specify object store and loggers via their yaml. I removed the CLI args and instead only instantiate object store/wandb if it's present in the yaml.

Reply:
Discussed offline -- keeping to control task specific logging and saving!